### PR TITLE
User typeahead enabled for non-admin project managers

### LINKED ISSUE #1152: Change add member dialog to search based on all users you can see
**Describe the feature**
User context:
> I have a project which has the user [user1], and another project that I would like to add him to. I tried adding him using his email address [assumed email], but it didn’t seem to find that user. (Maybe he used a different email address for his registration?) I had an idea that his Lexbox username was “[user1]”, so I tried that and it worked. But we really need a way to expose the user’s username from a project where I am manager, and that person is a member of the project.

We will change the "users in my org" query to also return users in projects you are part of (eg a manager of any project, or editor on non confidential projects), this will let users add other users they know of, without exposing their email.

**Who is this feature for?**
Project managers

**Pages affected**
Project page and permissions.

### PR DESCRIPTION
**Pull request by rmunn**
Fix #1152.

This PR adds a new `usersICanSee` GQL query, which looks for:

* All users in one of my orgs
* All users in one of the projects I manage
* All users in one of the **non-confidential** projects I'm a member of

The "Add Project Member" typeahead is now updated to use that query, which allows project managers who aren't site admins to use the typeahead to find users whose email address they don't know. E.g. if Test Manager has Test Editor as part of project A, and he also manages project B, then when he clicks on "Add Members" in project B, he can type Test Editor's name and select him to add, without knowing his email address.

### CODE REVIEW

**Comment by rmunn** [comment]
I've done local testing and the typeahead is working. I want to add some unit tests before calling this PR done, to make sure that the `usersICanSee` query is returning exactly the results it should. But the code is complete and working, so it's probably ready to review. The only changes I expect to make are:

- Add unit tests to ensure that `usersICanSee` returns what it should (e.g., no accidential disclosure of confidential project members)
- Perhaps delete the `usersInMyOrg` query if we don't expect to use it in the future.

File: backend/LexBoxApi/GraphQL/LexQueries.cs [hunk 1]
```diff
@@ -186,6 +186,22 @@ public IQueryable<User> UsersInMyOrg(LexBoxDbContext context, LoggedInContext lo
...
+    public IQueryable<User> UsersICanSee(LexBoxDbContext context, LoggedInContext loggedInContext)
+    {
+        var myOrgIds = loggedInContext.User.Orgs.Select(o => o.OrgId).ToList();
+        var myProjectIds = loggedInContext.User.Projects.Select(p => p.ProjectId).ToList();
+        var myManagedProjectIds = loggedInContext.User.Projects.Where(p => p.Role == ProjectRole.Manager).Select(p => p.ProjectId).ToList();
+        return context.Users.Where(u =>
+            u.Organizations.Any(orgMember => myOrgIds.Contains(orgMember.OrgId)) ||
+            u.Projects.Any(projMember =>
+                myManagedProjectIds.Contains(projMember.ProjectId) ||
+                (projMember.Project != null && projMember.Project.IsConfidential == false && myProjectIds.Contains(projMember.ProjectId))));
```
Comment (hahn-kev) [inline backend/LexBoxApi/GraphQL/LexQueries.cs:202]: I think we want `IsConfidential != true` because we decided to show users for unknown projects.

**Review by hahn-kev — CHANGES_REQUESTED** [review]
looks good, one minor change. For testing I would actually suggest moving the query code into `UsersService` and then calling it directly, that way you are not doing an e2e gql test, just a simple db query test, take a look at the `ProjectServiceTest` as an example which interacts with the database

File: backend/LexBoxApi/GraphQL/LexQueries.cs (hunk 1 above)
Comment (rmunn) [reply backend/LexBoxApi/GraphQL/LexQueries.cs]: Done in commit 44db0e32.

**Comment by rmunn** [comment]
VS Code told me that IEmailService was unused, so I removed it. Turns out it was used after all, just not in that file so the code analysis didn't spot it. Commit 2d0a7c56 should turn the unit test checks green again.

**Review by hahn-kev — CHANGES_REQUESTED** [review]
looks good, though I left a comment on the service tests that needs to be taken care of.

**Review by myieye — APPROVED** [review]
I agree with Kevin's request for the tests to be improved a bit, but otherwise this looks good to me 👍.

I pushed a few minor improvements:
- There were still a few things that wanted a rename
- Adapted the GenerateGqlSchema task/method so it keeps working
- Started hiding typeahead results that are duplicates of existing members (not quite specific to this PR, but relevant)

File: backend/Testing/LexCore/Services/UserServiceTest.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    [Fact]
+    public async Task NonManagerCanNotSeeUsersInConfidentialProjects()
+    {
+        var manager = await _lexBoxDbContext.Users.Include(u => u.Organizations).Include(u => u.Projects).FirstOrDefaultAsync(user => user.Email == "manager@test.com");
+        manager.ShouldNotBeNull();
+        var editor = await _lexBoxDbContext.Users.Include(u => u.Organizations).Include(u => u.Projects).FirstOrDefaultAsync(user => user.Email == "editor@test.com");
+        editor.ShouldNotBeNull();
+        await using var privateProject = await TempProjectWithoutRepo.Create(_lexBoxDbContext, true, manager.Id);
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/Testing/LexCore/Services/UserServiceTest.cs:61]: Yeah, that's not actually testing what I intended it to test until the editor is in the project. So what we want is some tests to verify the following:

* Admin can see everyone (which is a little pointless since they can already just query `users` and the frontend takes that into account, so maybe this test isn't necessary? It would require a code change anyway)
* Org members can see everyone in their org, and also anyone they share a public project with
* Project managers can see everyone in their projects, confidential or not, and also have the same rights as "normal" users in the projects they don't manage (can see public, can not see private)
* Non-managers can only see people they share public projects with (and org membership)

I think instead of using the users and projects from the seed data for this, I should have a test class constructor that sets up a specific set of projects and users, and a Dispose method on the test class that tears down all those projects and users. Then I can write the tests to the specific set of projects I created, without worrying about any changes I made to the seed data messing things up. (I once added some users to the sena-3 project and managed to make my tests start failing until I reset the data to the original seeded state. We don't want the tests to be that fragile).

File: frontend/src/lib/forms/UserTypeahead.svelte [hunk 3]
```diff
@@ -23,6 +24,8 @@
     [],
     debounceMs);
 
+  $: filteredResults = $typeaheadResults.filter(user => !exclude.includes(user.id));
```
Comment (rmunn) [inline frontend/src/lib/forms/UserTypeahead.svelte:27]: I like that this is in the frontend; it lets `usersICanSee` do exactly what it says on the <del>tin</del> function signature.

File: backend/Testing/ApiTests/UsersICanSeeQueryTests.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        MustHaveUser(json, "Qa Admin");
+    }
+
+    [Fact]
+    public async Task MemberCanSeeNotProjectMembersOfConfidentialProjects()
+    {
+        await LoginAs("manager");
+        await using var project = await this.RegisterProjectInLexBox(Utils.GetNewProjectConfig(isConfidential: true));
+        //refresh jwt
+        await LoginAs("manager");
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/Testing/ApiTests/UsersICanSeeQueryTests.cs:92]: I copied the logic from existing tests that were written before you tweaked the JWT refresh code. I did wonder whether that was necessary any more, but hadn't yet had time to check. I'll verify on Monday, then remove that line if it's no longer needed.

File: backend/Testing/ApiTests/UsersICanSeeQueryTests.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                    name
+                  }
+                }
+              }
+              """,
+            expectGqlError, expectSuccessCode: false);
+        return json;
+    }
+
+    private async Task AddUserToProject(Guid projectId, string username)
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/Testing/ApiTests/UsersICanSeeQueryTests.cs:28]: This class is going to be deleted in the final version of this PR; I just hadn't gotten around to deleting it yet.

File: backend/Testing/ApiTests/UsersICanSeeQueryTests.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                usersICanSee(take: 10) {
+                  totalCount
+                  items {
+                    id
+                    name
+                  }
+                }
+              }
+              """,
+            expectGqlError, expectSuccessCode: false);
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/Testing/ApiTests/UsersICanSeeQueryTests.cs:24]: Yes, that was leftover from my first attempt. But this class is going to be deleted anyway since this was the integration test suite that I've replaced with unit tests. (Which need a real database so they're not *exactly* "unit" tests, but close enough).

File: backend/Testing/LexCore/Services/UserServiceTest.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/Testing/LexCore/Services/UserServiceTest.cs]: Commit af794179 tests more scenarios for projects, but doesn't actually include org membership in the test. So I'll need to add a few more users in a couple different orgs to really make these tests complete.

**Comment by rmunn** [comment]
Commit 096d8258 adds a single failing test: a user who's a member of one private project but no orgs. You'd expect `usersICanSee` to return yourself, but currently it doesn't.

Right now that doesn't matter because the frontend filters out existing project members (including the member who can't see himself in the query), so he wouldn't be able to see himself in the frontend typeahead results even if the query was returning the right thing. But we probably *do* want the `usersICanSee` query to include yourself no matter what, otherwise it's not living up to its name. So I wrote a failing test for that edge case: Alan a Dale can't see himself in the `usersICanSee` query because he's not an org. (Little John *could* see himself because he could see the membership of the Outlaws organization, which includes him. Which is why Alan a Dale was *not* included in the Outlaws org, in order to exercise this edge case).

**Comment by rmunn** [comment]
@hahn-kev @myieye - I've rewritten the UserServiceTests pretty extensively, to set up a set of projects, orgs, and users used only for those tests. Would you let me know if the tests are clear and easy to understand, or if there's anything unclear about them? In particular, I'm hoping that the project and org memberships are pretty obvious when you're reading the tests, without having to scroll up to look at the InitializeAsync method to remind yourself of who is a member of which project. Did I achieve that goal?

**Review by hahn-kev — CHANGES_REQUESTED** [review]
looks good other than the new assertion package, lets remove it for now and discuss which one to use in our team meeting

**Comment by myieye** [comment]
LGTM 👍
So, I'll step away from this PR and let @hahn-kev approve when he's happy with the package choice.

File: backend/Testing/Testing.csproj [hunk 7]
```diff
@@ -27,6 +27,7 @@
         </Otherwise>
     </Choose>
     <ItemGroup>
+        <PackageReference Include="FluentAssertions" Version="6.12.0"/>
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/Testing/Testing.csproj:30]: Fluent assertions gave the nice convenient Should().BeEquivalentTo method which worked on enumerables, whereas the Shouldly assertions didn't have that. However now that I have my helper method I can probably rewrite it to use Shouldly.

[the remark this answers is not in the document]
Comment (hahn-kev) [reply backend/Testing/Testing.csproj:30]: Tim will merge in #1262 and Robin will merge this in after that

**Review by hahn-kev — APPROVED** [review]
